src/OpenLoco/StationManager.cpp

```diff
@@ -124,22 +128,210 @@ namespace OpenLoco::StationManager
+        {
+            // Forest
+            if ((realNamesInUse & (1 << 23)) == 0)
+                return StringManager::toTownName(StringIds::station_town_forest);
+        }
+        else if (numSurroundingTrees > 20)
+        {
+            // Woods
+            if ((realNamesInUse & (1 << 10)) == 0)
+                return StringManager::toTownName(StringIds::station_town_woods);
+        }
+
+        // 0x0048FB29
+        {
+            auto tile = TileManager::get(Map::Pos2(position.x, position.y));
+            for (auto& element : tile)
+            {
+                auto* surface = element.asSurface();
+                if (surface == nullptr)
+                    continue;
+
+                auto townHeightDiff = (position.z / 4) - surface->baseZ();
+                if (townHeightDiff > 20)
+                {
+                    // Heights
+                    if ((realNamesInUse & (1 << 9)) == 0)
+                        return StringManager::toTownName(StringIds::station_town_heights);
+                }
+                else if (townHeightDiff < -20)
+                {
+                    // Valley
+                    if ((realNamesInUse & (1 << 8)) == 0)
+                        return StringManager::toTownName(StringIds::station_town_valley);
+                }
+                break;
+            }
+        }
+
+        // 0x0048FB8B
+        if ((realNamesInUse & (1 << 0)) == 0)
+            return StringManager::toTownName(StringIds::station_town);
+
+        auto town = TownManager::get(townId);
+        {
+            auto xDiff = std::abs(position.x - town->x);
+            auto yDiff = std::abs(position.y - town->y);
+            if (xDiff + yDiff <= 120)
+            {
+                // Central
+                if ((realNamesInUse & (1 << 5)) == 0)
+                    return StringManager::toTownName(StringIds::station_town_central);
+            }
+        }
+
+        if (position.x <= town->x && position.y <= town->y)
+        {
+            // North
+            if ((realNamesInUse & (1 << 1)) == 0)
+                return StringManager::toTownName(StringIds::station_town_north);
+        }
+
+        if (position.x >= town->x && position.y >= town->y)
+        {
+            // South
+            if ((realNamesInUse & (1 << 2)) == 0)
+                return StringManager::toTownName(StringIds::station_town_south);
+        }
+
+        if (position.x <= town->x && position.y >= town->y)
+        {
+            // East
+            if ((realNamesInUse & (1 << 3)) == 0)
+                return StringManager::toTownName(StringIds::station_town_east);
+        }
+
+        if (position.x >= town->x && position.y <= town->y)
+        {
+            // West
+            if ((realNamesInUse & (1 << 4)) == 0)
+                return StringManager::toTownName(StringIds::station_town_west);
+        }
+
+        // 0x0048FC5C
+        string_id foundName = StringIds::empty;
+        for (auto i = 5; i < 47; i++)
```

Comment:
This looks a little horrible not sure really how it could be nicely tidied up though.
I think after naming all the constants you could do something like:
```
auto remainingRealNames = {StationRealNameFlags::sidings, StationRealNameFlags::annex, ...}
for (auto name : remainingRealNames){ ... }
```
Then repeat for ordinal names.